sympy/solvers/bivariate.py

```diff
@@ -181,7 +181,10 @@ def _lambert(eq, x):
     for arg in args:
         for k in lambert_real_branches:
             w = LambertW(arg, k)
-            if k and not w.is_real:
+            print(w.is_real, len((w).atoms(Symbol)), len((w).atoms(x)), w)
+            if (k and not w.is_real) and \
+                (not len((w).atoms(Symbol)) >= 1 \
+                    or len((w).atoms(x)) >= 1):
```

Comment:
We have to be careful about using `not` with `is_real` because `is_real` can give `None` and `not None` is `True`.

I am not sure I understand why this checks the number of symbols. Is that something to do with the case where `w.is_real` gives `None`?

Reply:
> We have to be careful about using not with is_real because is_real can give None and not None is True.
Yes, that's correct but the code was already present `(k and not w.is_real)`.(i.e. was not written by me)

> I am not sure I understand why this checks the number of symbols. Is that something to do with the case where w.is_real gives None

In a broad sense, yes. 
Actually, here expressions(`eq` in `def _lambert(eq, x):`) with symbols are considered as `None` so to compensate that the code `not len((w).atoms(Symbol)) >= 1`  and whereas for expressions (`eq` in `def _lambert(eq, x):`)  with multiple `x` (i.e. for example `solve([y, exp(x) + x])`) the code `or len((w).atoms(x)) >= 1` is written.

Reply:
What about specifying the symbol in categories like `a = Symbol('a',real=True)` , `a = Symbol('a',positive=True)` ?

Reply:
I think after the development of complex set infrastructure, the code will be more cleaner.

Reply:
> Yes, that's correct but the code was already present `(k and not w.is_real)`.(i.e. was not written by me)

I know it wasn't written by you but this code was part of `solve` and you are proposing to make it part of `solveset` in this PR. The requirements for `solve` and `solveset` are different so if we are going to make this part of `solveset` then we need to ensure that it is suitable for `solveset`.

As I said before if it is possible to improve this implementation so that both `solve` and `solveset` are better then that is great. On the other hand if this code can not be made suitable for `solveset` then it should not be used as part of `solveset`

> In a broad sense, yes.
> Actually, here expressions(`eq` in `def _lambert(eq, x):`) with symbols are considered as `None`(for `is_real`) so to compensate that the code
> 
> 1. `not len((w).atoms(Symbol)) >= 1`  and
> 2. whereas for expressions (`eq` in `def _lambert(eq, x):`)  with multiple `x`(in `def _lambert(eq, x):`) the code `or len((w).atoms(x)) >= 1` is written. ( for example `solve([y, exp(x) + x])` has multiple `x`)

I have no idea what your quoted explanation means. I feel like we're about to go around in circles with this like in previous discussions. The reason we go around in circles is because:

1. I feel sure that the proposed small piece of code is not correct or should be improved.
2. I don't understand enough about what the code is doing to say exactly what would be better code.
3. Each time I ask you about it you give an explanation that is not enough to convince me that the code is okay and is also not enough for me to be able to say what the correct code should be.
4. Then we go around in circles because points 1 and 2 still apply.

The situation here is that you know more than me about the general workings of solveset and the code that you are working with. However I know more than you about the general workings of sympy. Because I am not that familiar with the inner workings of solveset I am not going to merge any code that looks suspicious to me based on my broader knowledge of sympy.

Do not presume that I know more than you about this and also do not presume that you know more than me. This can only work if we can get to the point where we *both* understand the changes that are being made in the PR. I need you to help me understand what this code in solveset is doing and then I can help you to write the code in a better way.

Do not presume that I can follow everything that you have said in your previous comments though because there have been many and very few of them have given a clear explanation about anything: it is much better to write a single post with a concise and careful explanation that clarifies the point once than to write many posts that only partially explain things.

> I think after the development of complex set infrastructure, the code will be more cleaner.

This is not how it works! We can have a partial implementation of something in the codebase if it is not being used. If we want to make changes to a public function like `solveset` then each PR needs to make changes that only improve or maintain correctness (anything else is a bug). We do not merge a PR that is known to give incorrect results with the intention to fix the bugs in a later PR.

Reply:
I think we can't move further with the incomplete working of `is_real`.
And also the cases around `symbol` are made to compensate for that functionality(`is_real`).

Reply:
It might be possible to improve `is_real`:
https://github.com/sympy/sympy/blob/4b03990bb9df237c110b8120ffe5ed94c610348c/sympy/functions/elementary/exponential.py#L1145-L1163

Otherwise I don't think that the checking the number of symbols is a mathematically valid approach.

Reply:
I don't know the reasons for complications while dealing with `is_real` ( since may be due to specifications of a lot of different types of symbols and also may be due to dependency elsewhere)

What about this, here we are just allowing the negative branch (i.e. `LambertW( ..., -1)` ) for 
1.`symbols` and 
2. not allowing for `x`( `x` in `_lambert(eq, x)` ) ( e.g. in `solve([y, exp(x) + x])` here `x` is ( x, y ) which are not allowed since the `symbol` will be not present in solution --> solution : `{x: -LambertW(1), y: 0}`)

For which the code is...
```
if (k and not w.is_real) and \
      (not arg.has(Symbol) \
          or (arg).has(x)):
      continue
```

Reply:
I hope this might be a better approach : )

Reply:
So this code will remove solutions like `w = LambertW(arg, -1)` if:

1. `w.is_real` gives False or None
2. `arg` has no symbols
3. `arg` does not contain `x` (isn't `x` a symbol?)

The problem is that there will be some cases where `w.is_real` gives `None` when `w` is in fact real and does not have any symbols. This would remove those solutions.

Here's an example:
```python
In [90]: M = MatrixSymbol('M', 2, 2)

In [91]: w = LambertW(det(M), -1)

In [92]: print(w.is_real)
None

In [93]: w.atoms(Symbol)
Out[93]: set()
```

I think that it would be better just to change the condition here to `if w.is_real is False` and not check for symbols.

Reply:
```
>>> M = MatrixSymbol('M', 2, 2)
>>> w = LambertW(det(M), -1)
>>> solveset(w*exp(x) + x,x)
ConditionSet(x, Eq(x + exp(x)*LambertW(Determinant(M), -1), 0), Complexes)
>>> solveset(w*exp(x) + x,x,Reals)
Intersection(FiniteSet(-LambertW(LambertW(Determinant(M), -1))), Reals)

>>> arg = LambertW(Determinant(M), -1)
>>> arg.atoms(Symbol)
set()
>>> x = x

>>> arg.atoms(x)
set()
```

Reply:
```
>>> solveset(exp(x)+x, x,Reals)
1 arg
1 arg
FiniteSet(-LambertW(1))
>>> solve([y, exp(x) + x])
exp(_u) arg
exp(_u) arg
{y: 0, x: -LambertW(1)}

>>> from sympy.abc import x,y,z,a
>>> solveset(a*exp(x)+x, x,Reals)
a arg
a arg
Intersection(FiniteSet(-LambertW(a), -LambertW(a, -1)), Reals)


```

Reply:
> I think that it would be better just to change the condition here to if w.is_real is False and not check for symbols.

Then what about negative branches for solutions with symbols?
Are we going to compromise on that?

Reply:
Please do reply on this : )

Reply:
I don't understand what the comment means.

I have suggested changing it to `w.is_real is False`. Why doesn't that work?

Reply:
> I have suggested changing it to w.is_real is False. Why doesn't that work?

Instead of None ?

In that case only single branch will be returned !
Since if w.is_real is not real then there is no need for `-1` branch that is what the logic suggests !